crates/ruff_linter/src/rules/ruff/rules/implicit_optional.rs

```diff
@@ -202,11 +213,21 @@ pub(crate) fn implicit_optional(checker: &Checker, parameters: &Parameters) {
             else {
                 continue;
             };
-            let conversion_type = checker.target_version().into();
+
+            let mut add_future_import = false;
+            let conversion_type = if checker.target_version() >= PythonVersion::PY310 {
+                ConversionType::BinOpOr
+            } else if checker.settings().allow_importing_future_annotations {
+                add_future_import = true;
+                ConversionType::BinOpOr
+            } else {
+                ConversionType::Optional
+            };
 
             let mut diagnostic =
                 checker.report_diagnostic(ImplicitOptional { conversion_type }, expr.range());
-            diagnostic.try_set_fix(|| generate_fix(checker, conversion_type, expr));
+            diagnostic
+                .try_set_fix(|| generate_fix(checker, conversion_type, expr, add_future_import));
         }
```

Comment:
I think we can remove the `add_future_import` variable and instead compute it in `generate_fix`: It needs to add the future import if `BinOpOr` is `true` and the Python version is older than Python 3.10
```suggestion
            let mut add_future_import = false;
            let conversion_type = if checker.target_version() >= PythonVersion::PY310 || checker.settings().allow_importing_future_annotations {
                ConversionType::BinOpOr
            } else {
                ConversionType::Optional
            };

            let mut diagnostic =
                checker.report_diagnostic(ImplicitOptional { conversion_type }, expr.range());
            diagnostic
                .try_set_fix(|| generate_fix(checker, conversion_type, expr, add_future_import));
        }
```

Reply:
That's much nicer, thanks.